### ISSUE TITLE
XHR and resourceLoader

### ISSUE THREAD

**Issue by gustavnikolaj** [jsdom/jsdom#1393]
I was trying to use jsdom to facilitate server rendering of an application built with systemjs and react. I ended up doing that, as I needed to hook in with some existing tools for internationalization which were not built for running in the server.

My plan was to use jsdom to generate a window object and load the scripts, and then go load reacts renderToString method from that context. But I hit a problem when I realized that the custom resource loader is not used with XHR.

Is there a reason why the resource loader is not used in the XHR implementation?

**Comment by domenic** [comment]
Nope; sounds like a bug! PR with tests certainly welcome.

**Comment by kapouer** [comment]
Any update on this ? I'm willing to submit a PR, hopefully it's not that hard to setup.

**Comment by domenic** [comment]
No updates; PR welcome!

**Comment by kapouer** [comment]
Actually i only need to be able to hook into the "realClient" - maybe @gustavnikolaj too ? Replacing it asks for the custom resourceLoader to support so many things that it is bound to be useless.
Calling the custom resourceLoader with the realClient from xhr-utils would be just fine, allowing one to modify the client at will.

**Comment by kapouer** [comment]
After a few tries: it's not complicated to add that feature indeed, but it asks for a change regarding what  `defaultFetch` is supposed to return.
The most powerful solution is to return the current request, but that's not what was done in the currently implemented version for `download`.
That allows one to have full control over the properly setup default request - and in the case of xhr it's needed because it's difficult to build a custom xhr request without missing something.
Also if `defaultFetch` must return a request instance, then `responseLoader` should too.
Maybe `responseLoader` could be polymorphic: return request if no callback is specified.

**Comment by domenic** [comment]
I'd like to keep things backward compatible. However, if you have input into the new API resource loader, that would be good to consider.

**Comment by gagern** [comment]
Would it make sense to add a method `fetchXHR` or similar to the `ResourceLoader` class? And document that to be compatible with the single argument version of `request`, in that it must return an `EventEmitter` which uses `response` and the `stream.Readable` events.

Having a new method with a suitable default implementation would ensure perfect backwards compatibility, but at the same time offer implementers an extension point to tweak behavior. 

The code for handling the `file:` and `data:` protocols could be moved to the implementation of that method. For `http:` and `https:` protocols, the new method would merely delegate to `request` with no further processing. It could therefore serve as a drop-in replacement for all the `request` calls in `xhr-utils.js`.

**Comment by domenic** [comment]
No, I don't see a need for that architecture. Our current fetch hook should be sufficient, although it will take more work to adapt the XHR code to use it.

**Comment by gagern** [comment]
I'm just looking how other resources get fetched. https://github.com/jsdom/jsdom/blob/15.2.1/lib/jsdom/living/nodes/HTMLFrameElement-impl.js#L44-L45 for example makes use of the promise.request.headers property if available. Documenting that fact would go a long way towards explaining how custom loaders can provide headers. I was also surprised that `request-promise-native` does not really return a native Promise (according to `instanceof`), but more of a request-like thenable that will have a `response` property.

The XHR implementation would need to fall back to something sane if there is no `response` attribute on the promise. Where sane could mean playing it safe (i.e. not assuming any CORS permissions unless set explicitly) or relaxed (i.e. allowing anything and assuming that the custom loader will take care of required checks).

Incremental updates as the data is being loaded would be very hard with the promise-based approach, but might not be needed. I'm not sure how relevant realistic support for `onprogress` even handlers is for jsdom customers using this feature. On the other hand, with streams that would be kind of natural.

The [options object](https://github.com/jsdom/jsdom/blob/15.2.1/lib/jsdom/browser/resources/resource-loader.js#L58) would need expanding, but that shouldn't be too bad. Passing some [more headers](https://github.com/jsdom/jsdom/blob/15.2.1/lib/jsdom/living/xhr-utils.js#L234-L256) would be the most relevant change there, I think. Documenting possible options would be nice, too.

**Comment by domenic** [comment]
Hmm, yeah, the current resource loader is pretty tightly coupled to request-promise. We should indeed document that better. (Although it can be mostly behind the scenes, if you call `super.fetch()`.)

One day when we have a Fetch API implementation we could have it be a more service worker-like hook that takes a `Request` and returns a promise for a `Response`.

**Comment by gagern** [comment]
This “if you call `super.fetch()`” is a big if for me: I'm trying to write tests that simulate other host's behavior without actually contacting those hosts, so that the test does not depend on any external resources. I guess I might not be the only one with no intention of forwarding requests.

I believe that an event-based approach might fit the HTTP standard better than a promise-based one, in term of “one day” dreams.

One big simplification would be having an API for the object returned from the fetch. Then you could call methods on that object, such as “set headers” or “set body”, but also “transfer data chunk” and “end chunked data” as a streaming alternative to setting the whole body in one go. You could also document abort behavior in there. Having an API there would tell implementers what information they can provide, and how to do so. And you could have convenience constructors that construct such a response from a request, or from a promise for the body, or from static content, or whatever else you fancy.

**Comment by domenic** [comment]
I think we're pretty committed to the future path being to follow the fetch standard's Request/Response pair objects, instead of something event based or with the methods you describe.

**Comment by lafkpages** [comment]
Yeah, XMLHttpRequests don't get caught by the resource loader. I suppose I'll inject some JS to overwrite the constructor or something.
